Failed PUT on /imposters DELETES all imposters

#### Expected behaviour

If I submit a bad PUT request with valid JSON, for example an empty JSON object as the body "{}" to /imposters where imposters already exist I'd expect a 400 response with errors explaining to me the issue, but more importantly - the resource (imposters) to not be affected.

Request

```
PUT /imposters HTTP/1.1
Host: localhost:80
Content-Type: application/json
Cache-Control: no-cache

{}
```

Response

```
HTTP/1.1 400 Bad Request
Access-Control-Allow-Origin: *
Content-Type: application/json; charset=utf-8
Content-Length: 28
Date: Wed, 31 May 2017 03:11:53 GMT
Connection: keep-alive

{
  "errors": [
    {
      "code": "invalid request",
      "message": "Request body dun goofed",
      "source": "{}"
    {
  ]
}
```

Existing imposters unaffected

#### Actual behaviour

Response

```
HTTP/1.1 400 Bad Request
Access-Control-Allow-Origin: *
Content-Type: application/json; charset=utf-8
Content-Length: 28
Date: Wed, 31 May 2017 03:11:53 GMT
Connection: keep-alive

{
  "errors": [
    {}
  ]
}
```

All Imposters are deleted

#### Steps to reproduce

Create imposter 1
PUT /imposters with empty or invalid imposters call (but valid JSON)
View imposters
Imposters gone!


#### Software versions used

```
OS         : Windows x64
mountebank : 1.11.0
node.js    : N/A (x64 zip)
  (only if installed via npm)
Installation method : zip
```

#### Log contents in mb.log when running mb --loglevel debug

```
info: [mb:80] POST /imposters
info: [http:10001] Open for business...
info: [mb:80] PUT /imposters
info: [http:10001] Ciao for now
warn: [mb:80] error creating imposter: {"message":"Cannot read property 'map' of
 undefined","name":"TypeError","stack":"TypeError: Cannot read property 'map' of
 undefined\n    at C:\\Users\\XXX\\mountebank-v1.11.0-win-x64\\mountebank\\src
\\controllers\\impostersController.js:198:66\n    at _fulfilled (C:\\Users\\XXX
0\\mountebank-v1.11.0-win-x64\\mountebank\\node_modules\\q\\q.js:854:54)\n    at
 self.promiseDispatch.done (C:\\Users\\XXX\\mountebank-v1.11.0-win-x64\\mounte
bank\\node_modules\\q\\q.js:883:30)\n    at Promise.promise.promiseDispatch (C:\
\Users\\XXX\\mountebank-v1.11.0-win-x64\\mountebank\\node_modules\\q\\q.js:816
:13)\n    at C:\\Users\\XXX\\mountebank-v1.11.0-win-x64\\mountebank\\node_modu
les\\q\\q.js:624:44\n    at runSingle (C:\\Users\\XXX\\mountebank-v1.11.0-win-
x64\\mountebank\\node_modules\\q\\q.js:137:13)\n    at flush (C:\\Users\\XXX\\
mountebank-v1.11.0-win-x64\\mountebank\\node_modules\\q\\q.js:125:13)\n    at _c
ombinedTickCallback (internal/process/next_tick.js:73:7)\n    at process._tickDo
mainCallback (internal/process/next_tick.js:128:9)"}

```

The server returns `400 BAD REQUEST` when an empty JSON body is used because there is no `imposters` key which is assumed to be present by the controller. I could reproduce by adding a simple test: https://github.com/Mario54/mountebank/commit/e768eb5cae69669e31ab0d996e39f12fc5433c3f.

Here's the error message:
```
 1) ImpostersController #put should return an empty array if the request body is empty:

      AssertionError: { errors:
   [ TypeError: Cannot read property 'map' of undefined
       at /Users/mario/Code/mountebank/src/controllers/impost deepEqual { imposters: [] }
      + expected - actual

       {
      -  "errors": [
      -    "TypeError: Cannot read property 'map' of undefined"
      -  ]
      +  "imposters": []
       }

    at test/controllers/impostersControllerTest.js:293:24
    at _fulfilled (node_modules/q/q.js:854:54)
    at self.promiseDispatch.done (node_modules/q/q.js:883:30)
    at Promise.promise.promiseDispatch (node_modules/q/q.js:816:13)
    at node_modules/q/q.js:624:44
    at runSingle (node_modules/q/q.js:137:13)
    at flush (node_modules/q/q.js:125:13)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickDomainCallback (internal/process/next_tick.js:122:9)
```

However, this is not the core of issue. As this is a PUT request, specifying an empty JSON body could be equivalent to deleting all imposters. If this is the desired behaviour, the server should delete all the imposters and return `200`.

Another option would be to require the caller to include the `imposters` key if the desired behaviour is to delete all imposters.

```json
{
    "imposters": []
}
```

With this approach, an empty JSON body would cause the server to return `400` with an error explaining that the "`imposters` key is required", and the imposters would not be deleted.

@bbyars Which option would best align with the way mountebank was designed?

--- a/src/controllers/impostersController.js
+++ b/src/controllers/impostersController.js
@@ -191,6 +191,13 @@ function create (protocols, imposters, Imposter, logger) {
 
         logger.debug(requestDetails(request));
 
+        if (!('imposters' in request.body)) {
+            respondWithValidationErrors(response, [
+                exceptions.ValidationError("'imposters' is a required field")
+            ]);
+            return Q(false);
+        }
+
         return Q.all(validationPromises).then(function (validations) {
             var isValid = validations.every(function (validation) {
                 return validation.isValid;
@@ -198,7 +205,7 @@ function create (protocols, imposters, Imposter, logger) {
 
             if (isValid) {
                 return deleteAllImposters().then(function () {
-                    var creationPromises = request.body.imposters.map(function (imposter) {
+                    var creationPromises = requestImposters.map(function (imposter) {
                         return Imposter.create(protocols[imposter.protocol], imposter);
                     });
 

--- a/test/controllers/impostersControllerTest.js
+++ b/test/controllers/impostersControllerTest.js
@@ -285,12 +285,35 @@ describe('ImpostersController', function () {
             };
         });
 
+        promiseIt('should return a 400 if the "imposters" key is not present', function () {
+            var existingImposter = { stop: mock() },
+                imposters = { 0: existingImposter },
+                controller = Controller.create({ http: Protocol }, imposters, {}, logger);
+
+            request.body = {};
+
+            return controller.put(request, response).then(function () {
+                assert.strictEqual(response.statusCode, 400);
+                assert.deepEqual(response.body, {
+                    errors: [{
+                        code: 'bad data',
+                        message: "'imposters' is a required field"
+                    }]
+                });
+
+                assert.deepEqual(imposters, { 0: existingImposter });
+            });
+        });
+
         promiseIt('should return an empty array if no imposters provided', function () {
-            var controller = Controller.create({ http: Protocol }, {}, {}, logger);
+            var existingImposter = { stop: mock() },
+                imposters = { 0: existingImposter },
+                controller = Controller.create({ http: Protocol }, imposters, {}, logger);
             request.body = { imposters: [] };
 
             return controller.put(request, response).then(function () {
                 assert.deepEqual(response.body, { imposters: [] });
+                assert.deepEqual(imposters, {});
             });
         });
 

